File: Sort List/Sort List.py
# -*- coding: utf-8 -*-
#!/usr/bin/env python
# Sort a linked list in O(n log n) time using constant space complexity.
# Definition for singly-linked list.
import sys
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def sortList(self, head):
        if head == None:
            return None
        if head.next == None:
            return head

    def heap_build(self, x, heap, index):
        while heap[index/2] > x:
            heap[index] = heap[index/2]
            index = index/2
        heap[index] = x

    def delete_min(self, heap, size):
        if len(heap) == 0:
            return heap
        min_element = heap[1]
        last_element = heap[size]
        size = size - 1
        i = 1
        while i * 2 <= size:
            child = i * 2
            if child != size and heap[child+1] < heap[child]:
                child = child + 1
            if last_element > heap[child]:
                heap[i] = heap[child]
            else:
                i = child
                break
        heap[i] = last_element
        return min_element

s = Solution()
heap = [0 for i in range(1000)]
heap[0] = -sys.maxint
s.heap_build(21, heap, 1)
s.heap_build(13, heap, 2)
s.heap_build(16, heap, 3)
s.heap_build(19, heap, 4)
s.heap_build(14, heap, 5)
s.heap_build(31, heap, 6)
print heap[1:7]
# print s.delete_min(heap, 6)
# print s.delete_min(heap, 5)

File: Sort List/Sort List.cpp
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x): val(x),next(NULL){}
};

class Solution {
public:

	ListNode *sortList(ListNode *head) {	
		if(head == NULL || head->next == NULL) {
			return head;
		}
		ListNode *p = head, *q = head, *pre = NULL;
		while(q && q->next) {
			pre = p;
			p = p->next;
			q = q->next->next;
		}
		pre->next = NULL;				
		return mergeList(sortList(head), sortList(p));
	}

	ListNode *mergeList(ListNode *l, ListNode * r) {
		ListNode *p1, *p2, *result;
		if(l->val < r->val) {
			p1 = l;
			p2 = r;			
		} else {
			p1 = r;
			p2 = l;

		}
		result = p1;
		ListNode * pre;
		while(p1 && p2) {
			if(p1->val <= p2->val) {
				pre = p1;
				p1 = p1->next;
			} else {
				pre->next = p2;
				p2 = p2->next;
				pre = pre->next;
				pre->next = p1;
			}
		}
		if(p2) {
			pre->next = p2;
		}
		return result;
	}
	

};

ListNode *createList(int a[], int n) {
	ListNode *head = NULL, *p=NULL;
	for(int i = 0; i < n; i++) {
		if(head == NULL) {
			head = p = new ListNode(a[i]);
		} else {
			p->next = new ListNode(a[i]);
			p = p->next;
		}
	}
	return head;
}

void printList(ListNode *a) {
	while(a!=NULL) {
		printf("%d ", a->val);
		a = a->next;
	}
	printf("\n");
}

int main() {
	srand(time(NULL));
	int *a = new int[10];
	for(int i = 0; i<10; i++) {
		a[i] =  rand()%10+1;
	}
	Solution s;
	ListNode *p = createList(a, 10); 
	printList(p);
	ListNode *end;
	ListNode *q = s.sortList(p);
	printList(q);
	int b = 2, c = 3;
	printf("%d, %d", b, c);

}